### ISSUE TITLE
Parent processes (winpty-agent.exe, etc.) does not exit when shell process exits by itself

### ISSUE THREAD

**Issue by Luluno01** [microsoft/node-pty#333]
## Environment details

- OS: Windows 10 Pro 64bit
- OS version: Windows 10 Pro 64bit Version 1809 (OS Build 17763.615)
- node-pty version: 0.9.0-beta19

## Issue description

If a shell process suicides, its parent processes (not including the main node.js process) will not exit automatically unless `pty.kill()` is called. Check the following issue reproducing code:

```
import { spawn } from 'node-pty'
import * as os from 'os'


function getUserHome() {
  return process.env[process.platform == 'win32' ? 'USERPROFILE' : 'HOME']
}

function getShell() {
  switch (os.platform()) {
    case 'win32': return process.env.ComSpec || 'cmd.exe'
    case 'android': return '/bin/sh'
    default: return '/bin/bash'
  }
}

function createProcess({ cols, rows }: { cols?: number, rows?: number } = { cols: 80, rows: 30 }) {
  const shell = getShell()
  return spawn(shell, [], {
    name: 'xterm-color',
    cols: cols || 80,
    rows: rows || 30,
    cwd: getUserHome(),
    env: process.env as { [key: string]: string },
    experimentalUseConpty: true
  })
}

function sleep(n): Promise<void> {
  return new Promise(res => setTimeout(res, n))
}

async function main() {
  await sleep(10 * 1000)
  for(let i = 0; i < 10; i++) {
    const pty = createProcess()
    console.log(pty.pid)
    await sleep(1000)
    pty.write('exit\r')
  }
  await sleep(60 * 1000)
}

main()

```

The code above creates 10 `winpty-agent.exe`s and 10 `conhost.exe`. However, `node-pty` has notified user that those shell processes has exited by emitting `exit` event, misleading users that the process has exited cleanly.

It seems that there is no graceful way to determine whether the process is killed by calling `pty.kill()` or exited by itself (and therefore we should call `pty.kill()` inside the `onExit` handler). However, invoking `pty.kill()` twice on the same process will raise error (can be caught though).

**Comment by ViceIce** [comment]
Seeing this since i use vscode [remote dev](https://code.visualstudio.com/docs/remote/remote-overview) feature using docker container:

![image](https://user-images.githubusercontent.com/1798109/64149564-76e04880-ce26-11e9-89e2-5b56baa8a821.png)

**Comment by gpetrov** [comment]
@Tyriar any news in addressing this issue? We end up with hundreds of winpty-agent.exe processes ...

**Comment by Luluno01** [comment]
> @Tyriar any news in addressing this issue? We end up with hundreds of winpty-agent.exe processes ...

@gpetrov Are you using `node-pty` itself or VSCode? If you are using `node-pty` and have control on the PTY object, you can always call `pty.kill()` in `onExit` handler (note that `pty.kill` will throw an error if the group of processes has already been killed).

**Comment by gpetrov** [comment]
@Luluno01 Yes I went indeed for pty.kill() onExit for now - this helps indeed. It just didn't seem like a neat solution but it does the job for now.

And no we have a different project using node-pty and not VSCode

**Comment by Tyriar** [comment]
winpty support is legacy at this point and is only sticking around until the conpty dll code is ready to turn on by default.

**Comment by daguej** [comment]
I'm seeing this in 1.1.0-beta34 (Win 11 26100.3775) while using conpty, and worse, the orphaned conhost.exe does not exit even after manually calling `pty.kill()` in `onExit`.

```ts
import { spawn } from 'node-pty';
const pty = spawn(cmd, argv, {
  useConpty: true
});
pty.onExit(e => {
  console.log('exit ' + e.exitCode);
  try {
    pty.kill();
  } catch (e) { console.error(e) }
});

```

If the spawned `cmd` exits on its own, I'll see `exit 0` and no errors logged, but the conhost.exe process never exits.

If I call `pty.kill()` before `cmd` exits, both the spawned process and associated conhost.exe exit.

**Comment by Tyriar** [comment]
We'll be removing winpty soon https://github.com/microsoft/node-pty/issues/842
